### PR TITLE
Completely remove DispatchContext

### PR DESCRIPTION
**Pull request by iamed2**
I figure this is still okay to do in one shot since we're pre-1.0.

@rofinn please review when you can

@nicoleepp please let me know if you were relying on anything I removed

### CODE REVIEW

**Review by nicoleepp — APPROVED** [review]
Doesn't affect anything being relied on for `DaskDistributedDispatcher.jl`.

File: src/executors.jl [hunk 1]
```diff
@@ -49,137 +49,138 @@ and should be retried (and `false` otherwise). The default `retry_on` method ret
...
-Return the first argument.
+Construct a graph starting from `input_nodes` and ending in `output_nodes`.
+The graph is created by recursively identifying dependencies of nodes starting with
+`output_nodes` and ending with `input_nodes` (dependencies of `input_nodes` are not added to
+the graph).
 """
-function add_cleanup_nodes!(
-    ctx::DispatchContext;
-    exclude::Vector=DispatchNode[],
+function build_graph{T<:DispatchNode, S<:DispatchNode}(
```
Comment (rofinn) [inline src/executors.jl]: Could these just be `DispatchGraph` constructors to minimize cluttering up the namespace?

File: test/runtests.jl [hunk 2]
```diff
@@ -838,78 +759,72 @@ end
...
                         end
                     )
-                    set_label!(op, "3")
-                    @test isempty(dependencies(op))
-                    a = add!(ctx, op)
+                    set_label!(a, "3")
+                    @test isempty(dependencies(a))
 
-                    op = Op(
+                    b = Op(
```
Comment (rofinn) [inline test/runtests.jl]: Might be cleaner to write these as do-blocks.

File: src/executors.jl [hunk 3]
```diff
@@ -251,7 +252,7 @@ failing_node = add!(ctx, Op(()->throw(ErrorException("ApplicationError"))))
 dep_node = add!(n -> println(n), failing_node)  # This will fail as well
 ```
 
-Then `dispatch!(exec, ctx)` will throw a `DependencyError` and
+Then `dispatch!(exec, ctx.graph)` will throw a `DependencyError` and
```
Comment (rofinn) [inline src/executors.jl]: `ctx.graph` -> `graph`.

File: src/executors.jl [hunk 4]
```diff
@@ -251,7 +252,7 @@ failing_node = add!(ctx, Op(()->throw(ErrorException("ApplicationError"))))
...
 `dispatch!(exec, ctx; throw_error=false)` will return an array of passing nodes and the
```
Comment (rofinn) [inline src/executors.jl]: `ctx` -> `graph`

File: src/executors.jl [hunk 5]
```diff
@@ -268,7 +269,7 @@ http_node = add!(ctx, Op(()->http_get(...)))
 ```
 
 Assuming that the `http_get` function does not error 5 times the call to
-`dispatch!(exec, ctx)` will return [n1, n2, http_node].
+`dispatch!(exec, ctx.graph)` will return [n1, n2, http_node].
```
Comment (rofinn) [inline src/executors.jl]: `ctx.graph` -> `graph`

File: docs/src/pages/api.md [hunk 6]
```diff
@@ -86,34 +87,15 @@ add_edge!(::DispatchGraph, ::DispatchNode, ::DispatchNode)
...
 
 ```@docs
 Executor
-run!{T<:DispatchNode, S<:DispatchNode}(exec::Executor, ctx::DispatchContext, nodes::AbstractArray{T}, input_nodes::AbstractArray{S})
-run!(::Executor, ::DispatchContext)
+run!{T<:DispatchNode, S<:DispatchNode}(exec::Executor, nodes::AbstractArray{T}, input_nodes::AbstractArray{S})
+run!(::Executor, ::DispatchGraph)
+build_graph
 prepare!(::Executor, ::DispatchContext)
 dispatch!(::Executor, ::DispatchContext)
```
Comment (rofinn) [inline docs/src/pages/api.md]: `DispatchContext` -> `DispatchGraph`

**Review by rofinn — CHANGES_REQUESTED** [review]
This seems a lot cleaner.

**Review by rofinn — APPROVED** [review]

**Pull request merged**